src/poetry/repositories/http_repository.py

```diff
@@ -98,10 +99,29 @@ def _get_info_from_sdist(self, url: str) -> PackageInfo:
         with self._cached_or_downloaded_file(Link(url)) as filepath:
             return PackageInfo.from_sdist(filepath)
 
-    def _get_info_from_urls(self, urls: dict[str, list[str]]) -> PackageInfo:
+    @staticmethod
+    def _get_info_from_metadata(
+        url: str, metadata: dict[str, pkginfo.Distribution]
+    ) -> PackageInfo | None:
+        if url in metadata:
+            dist = metadata[url]
+            return PackageInfo(
+                name=dist.name,
+                version=dist.version,
+                summary=dist.summary,
+                requires_dist=list(dist.requires_dist),
+                requires_python=dist.requires_python,
+            )
+        return None
+
+    def _get_info_from_urls(
+        self,
+        urls: dict[str, list[str]],
+        metadata: dict[str, pkginfo.Distribution] | None = None,
+    ) -> PackageInfo:
+        metadata = metadata or {}
         # Prefer to read data from wheels: this is faster and more reliable
-        wheels = urls.get("bdist_wheel")
-        if wheels:
+        if wheels := urls.get("bdist_wheel"):
```

Comment:
As far as I got it, we risk falling back to package inspection when:
- A high priority package (wheel) does not have PEP 658 metadata
- A low priority package (sdist) does have PEP 658 metadata

In this case we would get the sdist metadata from the repository but then we wouldn't use it.

I would prefer a more aggressive strategy that favours PEP 658 over package inspection. Something like (in this order):
1. PEP 658 metadata of wheel
2. PEP 658 metadata of sdist
3. wheel inspection
4. sdist inspection

While currently it feels more like:
1. PEP 658 metadata of wheel
2. wheel inspection
3. PEP 658 of sdist
4. sdist inspection

One way you could achieve this is by reverting changes to `_get_info_from_urls` (which would still act on packages only) and inside `_links_to_data` something like:
```python
if metadata:
    # We got PEP 658 metadata from at least one package.
    # Either take one randomly or use some fancier logic (wheels over sdist)
    info = _get_info_from_metadata(metadata)
else:
    info = _get_info_from_urls(urls)
```

Reply:
Possible network optimization in case we go for the "random" approach: as soon as we got PEP 658 metadata from one url, stop querying others as we don't need anything else

Reply:
I think we can keep it as is for now. The current approach has the advantage that it does not change our logic (which wheel's metadata we chose) so we get the same results no matter if we can download metadata or if we have to download the wheel/sdist.

Further, I suppose that in 99 % of use cases it's not relevant because either all files of a release have PEP 658 metadata or none of them. (I may be wrong, but in that case we can adjust it later.)